cloud_info_provider/core.py

```diff
@@ -1,180 +1,32 @@
-        site_info = self._get_info_from_providers('get_site_info')
-
-        # Get shares / projects and related images and templates
-        shares = self._get_info_from_providers('get_compute_shares')
-
-        for share in shares.values():
-            kwargs = share.copy()
-
-            endpoints = self._get_info_from_providers('get_compute_endpoints',
-                                                      **kwargs)
-            if not endpoints.get('endpoints'):
-                return ''
+import logging
 
-            # Collect static information for endpoints
-            static_compute_info = dict(endpoints, **site_info)
-            static_compute_info.pop('endpoints')
-
-            # Collect dynamic information
-            images = self._get_info_from_providers('get_images',
-                                                   **kwargs)
-            templates = self._get_info_from_providers('get_templates',
-                                                      **kwargs)
-            instances = self._get_info_from_providers('get_instances',
-                                                      **kwargs)
-            quotas = self._get_info_from_providers('get_compute_quotas',
-                                                   **kwargs)
-
-            # Add same static information to endpoints, images and templates
-            for d in itertools.chain(endpoints['endpoints'].values(),
-                                     templates.values(),
-                                     images.values()):
-                d.update(static_compute_info)
+from stevedore import driver
+from stevedore import extension
 
-            share['images'] = images
-            share['templates'] = templates
-            share['instances'] = instances
-            share['endpoints'] = endpoints
-            share['quotas'] = quotas
 
-        # XXX Avoid creating a new list
-        endpoints = {endpoint_id: endpoint for share_id, share in
-                     shares.items() for endpoint_id,
-                     endpoint in share['endpoints'].items()}
+def get_providers():
+    def _handle_exception(*args):
+        mgr, entry_point, exception = args
+        logging.getLogger('stevedore.extension').error((
+            "Cannot load '%s': %s") % (entry_point, exception))
 
-        # XXX Avoid redoing what was done in the previous shares loop
-        static_compute_info = dict(endpoints, **site_info)
-        static_compute_info.pop('endpoints')
-
-        info.update({'static_compute_info': static_compute_info})
-        info.update({'shares': shares})
-
-        return self._format_template('compute', info)
-
-
-class CloudBDII(BaseBDII):
-    def __init__(self, opts):
-        super(CloudBDII, self).__init__(opts)
-        self.templates = ('headers', 'clouddomain')
-
-    def render(self):
-        output = []
-        info = self._get_info_from_providers('get_site_info')
+    mgr = extension.ExtensionManager(
```

Comment:
Will this fail if the module cannot be imported (e.g. for an OpenNebula site where the OS clients may not be installed)?

Reply:
No, `_handle_exception()` method will take care of managing any exception that might be raised. As it is now it will report (as error) the exception and continue the execution.

Reply:
great!